### PR TITLE
Fixes #178: Add isEqualToComparingFieldByFieldRecursively

### LINKED ISSUE #178: Feature request: isEqualToComparingFieldByField - recursive
I like the idea of isEqualToComparingFieldByField, but it is of limited use when some of the fields are objects without Equals being overriden. 

I would like to have an option to use isEqualToComparingFieldByField recursively, instead of falling back to org.assertj.core.util.Objects.areEqual.

### PR DESCRIPTION
**Pull request by PascalSchumacher**

### CODE REVIEW

File: src/main/java/org/assertj/core/error/ShouldBeEqualByComparingFieldByFieldRecursive.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    return new ShouldBeEqualByComparingFieldByFieldRecursive("Expecting:  <%s>%nto equal to:  <%s>%nwhen recursively comparing field by field, but found the following difference(s):%n"
+                                                             + join(descriptionOfDifferences, format("%n")), actual, other);
+  }
+
+  private ShouldBeEqualByComparingFieldByFieldRecursive(String message, Object actual, Object other) {
+    super(message, actual, other);
+  }
+
+  private static String join(List<String> list, String conjunction) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldBeEqualByComparingFieldByFieldRecursive.java]: you can use `org.assertj.core.util.Strings.join` like this : `Strings.join("a", "b", "c").with("|");` instead.

File: src/test/java/org/assertj/core/internal/objects/Objects_assertIsEqualToComparingFieldByFieldRecursive_Test.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+import org.assertj.core.api.AssertionInfo;
+import org.assertj.core.internal.ObjectsBaseTest;
+import org.assertj.core.util.DeepDifference.Difference;
+import org.junit.Test;
+
+
+public class Objects_assertIsEqualToComparingFieldByFieldRecursive_Test extends ObjectsBaseTest {
+
+  @Test
+  public void should_be_able_to_compare_objects_recursivly() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/objects/Objects_assertIsEqualToComparingFieldByFieldRecursive_Test.java]: typo : `recursivly`

File: src/main/java/org/assertj/core/error/ShouldBeEqualByComparingFieldByFieldRecursive.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  public static ErrorMessageFactory shouldBeEqualByComparingFieldByFieldRecursive(Object actual, Object other,
+                                                                                  List<Difference> differences) {
+    List<String> descriptionOfDifferences = new ArrayList<>(differences.size());
+    for (Difference difference : differences) {
+      descriptionOfDifferences.add(format("%nPath to difference:  <%s>%nexpected:  <%s>%nbut was:  <%s>",
+                                          join(difference.getPath(), "."), difference.getOther(),
+                                          difference.getActual()));
+    }
+    return new ShouldBeEqualByComparingFieldByFieldRecursive("Expecting:  <%s>%nto equal to:  <%s>%nwhen recursively comparing field by field, but found the following difference(s):%n"
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldBeEqualByComparingFieldByFieldRecursive.java]: `to be equal to`

**Comment by PascalSchumacher** [comment]
@joel-costigliola Thanks for the feedback. I have updated the pull request.

**Comment by PascalSchumacher** [comment]
@joel-costigliola I added another commit with replaces `getDeepDeclaredFields` with `Objects#getDeclaredFieldsIncludingInherited`. This make this pull request closer aligned with the rest of AssertJ reduces amount of code to review.

**Comment by joel-costigliola** [comment]
cool, I will play with this we.

**Comment by PascalSchumacher** [comment]
Thanks, take your time, no need to hurry. :smile: 

I did some more clean-ups and squashed the commits.

**Comment by joel-costigliola** [comment]
@PascalSchumacher  I'm currently adding support for per type/field comparator, I'm hoping to finish it this week.

**Comment by PascalSchumacher** [comment]
great, I won't do any updates of the pull request in meanwhile.

**Comment by joel-costigliola** [comment]
@PascalSchumacher I'm thinking to set some default Double and Float comparators with a given precision that users would be able to override, that would impact existing field by field comparison, any objections ?

**Comment by PascalSchumacher** [comment]
@joel-costigliola No objections. Feel free to change anything.

**Comment by joel-costigliola** [comment]
Done, thanks for the work Pascal !

**Comment by PascalSchumacher** [comment]
Thanks for merging. I really like your additions. :)

**Pull request closed without merging**